### PR TITLE
Correctly handle device selection in `Image` and `ImageEditor`

### PR DESCRIPTION
**Pull request by hannahblair**
## Description

We weren't correctly handling the event that the video input device changes in the `Image` and `ImageEditor` component - this PR fixes that. 

Note, I'm struggling to find a way to test this and simulate multiple input devices in Playwright (or other). If anyone knows a way please let me know!

To test, use any external video device (if you have an iPhone you can seamlessly use the [continuity camera](https://support.apple.com/en-gb/102546)) and run the `demo/webcam` demo. Test that the video inputs alternate as expected. 

Closes: #7493

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by dawoodkhan82** [comment]
Could not get continuity camera to work, so couldn't thoroughly test. Also noticed on safari, the dropdown for device selection is empty.

Screenshot 2024-03-20 at 1 20 43 PM

**Comment by dawoodkhan82** [comment]
> you might have to update your MacOS, Chrome or iOS.

probably all of the above lol

**Review by dawoodkhan82 — COMMENTED** [review]
@hannahblair The device source list works on safari now, thanks for fixing.

**Comment by dawoodkhan82** [comment]
Another thing I noticed is that the aspect ratio of the webcam video is sometimes long or too short. Specifically on safari, but not sure if it's just me.

Screenshot 2024-03-21 at 2 15 24 PM

**Comment by pngwn** [comment]
This isn't quite working for me. 

I tested on macOS, chrome with two video inputs:

- macbook pro webcam
- iphone

- By default it was selecting the iphone camera but when i clicked the selection box the 'selected' input was the macbook pro webcam. 
- When i selected the iphone (changed the input), it 'switched' to the same device (the iphone camera).
- After this it worked correctly, switching to the mbp webcam changed to the correct input.

I also don't like that I have to click twice in order to get the dropdown to show up, i don't think it was like this originally. When I click the dropdown arrow, I expected to be able to select a new input straight away. (this can wait for a different PR tho).

File: js/image/shared/Webcam.svelte [hunk 1]
```diff
@@ -33,24 +35,40 @@
...
+			get_video_stream(include_audio, video_source)
+				.then(async (local_stream) => {
+					webcam_accessed = true;
+					available_video_devices = await get_devices();
+					stream = local_stream;
+				})
+				.then(() => set_available_devices(available_video_devices))
+				.then((devices) => {
+					available_video_devices = devices;
+					selected_device = devices[0];
```
Comment (pngwn) [inline js/image/shared/Webcam.svelte:66]:
```js
```js
selected_device =
  devices.find(
    (device) => device.label === stream.getVideoTracks()[0].label
  ) || available_video_devices[0];
```

Comment (pngwn) [reply js/image/shared/Webcam.svelte:66]: Or this as you found: https://stackoverflow.com/questions/66048083/enumeratedevices-after-getusermedia-how-to-find-the-active-devices/66072112#66072112

Comment (pngwn) [reply js/image/shared/Webcam.svelte:66]: Yeah, this is better:

```js
const used_devices = stream
  .getTracks()
  .map((track) => track.getSettings()?.deviceId)[0];

selected_device = used_devices
  ? devices.find(
      (device) => device.deviceId === (used_devices as string)
    ) || available_video_devices[0]
  : available_video_devices[0];
```

**Review by pngwn — APPROVED** [review]
Awesome! Thanks for taking care of this and following up those weird bugs @hannahblair!

**Pull request merged**